### ISSUE TITLE
Move team member flags to team table + join table

### ISSUE THREAD

**Issue by timhabermaas** [cubing/worldcubeassociation.org#185]
```
  `wdc_team` tinyint(1) DEFAULT NULL,
  `wdc_team_leader` tinyint(1) DEFAULT NULL,
  `wrc_team` tinyint(1) DEFAULT NULL,
  `wrc_team_leader` tinyint(1) DEFAULT NULL,
  `results_team_leader` tinyint(1) DEFAULT NULL,
  `wca_website_team_leader` tinyint(1) DEFAULT NULL,
  `software_admin_team` tinyint(1) DEFAULT NULL,
  `software_admin_team_leader` tinyint(1) DEFAULT NULL,
```

[This is getting pretty ridiculous](https://github.com/cubing/worldcubeassociation.org/commit/73633700e04310679f4e2b044b3e6797bc6320b8#commitcomment-13553134). A better solution might be to create two separate tables:

* A `teams` table with the columns `(name, description)`
* A join table between `teams` and `users` describing which user belongs to which team and whether he is a leader of that team. Columns: `(team_id, user_id, leader)`

Advantages:
* DRY way of enforcing common validation rules/db constraints
* No need to explicitly list the different teams when iterating over them
* Adding/removing users to/from different teams can be handled in a generic way (`INSERT`/`DELETE` in SQL)

Disadvantages:
* Getting the user's permissions requires a join
* This might introduce coupling between data and code (e.g. giving different teams different permissions requires mapping from the `teams.name` column to some function in Ruby. This could be solved by a (more or less) generic `permissions(team_id, may_post, may_create_competition, ...)` table, though. Not sure if introducing one is worth it.

**Comment by jfly** [comment]
If we do this, it would be cool to keep track of *when* people held their role, rather than deleting rows.

**Comment by pedrosino** [comment]
Maybe also have delegates separated like this.

**Comment by jfly** [comment]
Ah yes, I didn't realize this issue was only about team members, but we should definitely do this for delegates, and consequently, candidate delegates, senior delegates, and board members as well.

**Comment by pedrosino** [comment]
So, I started tracking when people became delegates (only went back to november so far :P)

I'll try to list the tasks needed for this to become reality:
- [ ] one table with team's basic info (id, name, description?, leader?)
- [ ] one table joining team(s) and user(s) (team_id, user_id, start_date, end_date)
- [ ] update the user/edit page/action to work with the new structure
- [ ] migrate all data from users table to the new tables
- [ ] dig into the past and fill the new tables - maybe for too old stuff we could use the first competition they delegated as start_date, and the last competition as end_date.

**Comment by jfly** [comment]
:+1: :+1: :smile: :grinning: :fireworks: :fire:

**Comment by timhabermaas** [comment]
> one table joining team(s) and user(s) (team_id, user_id, start_date, end_date)

I like this design. `end_date = NULL` means the user is currently a delegate and deleting entries would translate to `UPDATE end_date=timestamp()`, right? I think I came across that idea on [Udi Dahan's blog](http://udidahan.com/?blog=true), but I can't find the post anymore.

**Comment by pedrosino** [comment]
That sounds right :)

**Comment by pedrosino** [comment]
One thing, though. I'm not very excited to search for when people became results or wdc members...I guess we could just set the start_date to today and hope we don't need that past information.

**Comment by pedrosino** [comment]
I'm a little puzzled here, maybe someone can help.

A team has a leader, which is a user. What is the correct way to implement this? I initially thought it would be like
```
create_table :teams do |t|
      t.string :friendly_id
      t.string :name
      t.text :description
      t.integer :leader

      t.timestamps null: false
    end
```
But that will not automatically associate a user to that number. I found http://guides.rubyonrails.org/association_basics.html#the-has-one-association but I'm having trouble on how to declare it. This is what I have:

```
class Team < ActiveRecord::Base
  has_many :users, through: :team_member
end
```

I thought of creating a team_leader table, but that sounds unnecessary, since each team only has one leader.

Bottom line, how do I include a leader in the Team model and have it as a User?

**Comment by timhabermaas** [comment]
First of all, I'd use ActiveRecord's convention and add `_id` at the end of the `leader` column:

```Ruby
create_table :teams do |t|
  t.string :friendly_id
  t.string :name
  t.text :description
  t.integer :leader_id

  t.timestamps null: false
end
```

Then in order to get the association working you need to do the following:

```Ruby
class Team < ActiveRecord::Base
  belongs_to :leader, class_name: "User"
end
```

Simply `belongs_to :leader` won't cut it since ActiveRecord will look for a class named `Leader` which doesn't exist. So you need to point it to the correct class which is `User`.

Unfortunately `belongs_to` sounds kinda strange in this case, but it simply means "I have a foreign key called x" while `has_one` means "the other table has a foreign key called y".

**Comment by pedrosino** [comment]
Thanks @timhabermaas 
I figured it out after searching and trying. Indeed the belongs_to sounds weird here. We don't want the team leaders to think they own the team :P

I thought of having members inside of the teams/edit view, similar to the polls and poll_options. Not sure it's the best idea.
It seemed to work at first, but it is not updating team_members. It's inserting new (repeated) records everytime...

**Comment by timhabermaas** [comment]
Are you using `accepts_nested_attributes_for`? If so, have you added `allow_destroy: true`?

**Comment by pedrosino** [comment]
Yeah, I forgot to permit :id in the team_member_attributes, so it was not updating.
Just made #390 
Please let me know if there's something weird.

**Comment by pedrosino** [comment]
I just ran
```
select id, name, wdc_team, wdc_team_leader, wrc_team, wrc_team_leader, software_team, software_team_leader, results_team, results_team_leader
from users
where wdc_team <> '0' OR wdc_team_leader <> '0' OR wrc_team <> '0' OR wrc_team_leader <> '0'  OR software_team <> '0'  OR software_team_leader <> '0' OR results_team <> '0' OR results_team_leader <> '0'
order by wdc_team, wdc_team_leader, wrc_team, wrc_team_leader, software_team, software_team_leader
```
And found 20 people:
![2016-02-19-085649_609x387_scrot](https://cloud.githubusercontent.com/assets/1881933/13173810/c374aa0a-d6e6-11e5-8bc5-5091c2fded35.png)

Doesn't seem to hard to migrate data :)
But there are some inconsistencies, like some people have NULL while others have 0, and @Luis-J-Ianez has an abbreviation in his name!

**Comment by pedrosino** [comment]
To ellaborate a bit on the mentioned thing about people having NULL or 0, just an example:
```
select wdc_team, count(*) from users
group by wdc_team
```

```
NULL   11262
0        435
1          3
```

```
select wrc_team, count(*) from users
group by wrc_team
```

```
NULL   11262
0        433
1          5
```

All those with 0 were updated at 2015-11-24 or sooner, but some of those with NULL were updated as recently as today.
Maybe it has something to do with admins updating users x users updating themselves...
But hopefully we'll get rid of these columns soon, so it shouldn't matter anymore, unless @jfly wants to investigate further.
